### PR TITLE
DOC GH22897 Fix docstring of join in pandas/core/frame.py

### LINKED ISSUE #22897: DOC: Fix the docstring of join in pandas/core/frame.py
pandas docstring are expected to follow the format defined in https://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html

This includes among other things:
- There is a short description fitting in one line (there can be a longer one after that).
- The description of the parameters, the returns section, and the see also items, starts with a capital letter and ends with a period.
- In the Return section there is no name before the type (e.g. `transformed : Series`) but just the type (e.g. `Series`) in the first line.
- The examples are valid PEP-8 code, and when executed, they produce the shown input.

We should make the required adjustments in the docstring of the function `join` in `pandas/core/frame.py`.

We should validate the docstring with our validator (which does not validate all the formats, but some):
- `./scripts/validate_docstrings.py pandas.DataFrame.join`

And we should validate the PEP-8 of the examples (note that the next command will validate the PEP-8 of all examples, just make sure that none of the shown is in the function being addressed, the rest will be fixed in separate issues):
- `flake8 --doctests pandas/core/frame.py`

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [X] closes #22897
- [X] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

#22900 was dealt with by adding an "import pandas as pd" line to the start of the examples section.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -6440,6 +6440,8 @@ def append(self, other, ignore_index=False,
     def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
              sort=False):
         """
+        Append columns of another DataFrame.
```
Comment (datapythonista) [inline pandas/core/frame.py]: Can you replace `append` by `join`? I think in general they mean the same, but in this context may give the idea that the they are being added at the end of the DataFrame, or that they are not aligned.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -6449,31 +6451,31 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
         other : DataFrame, Series with name field set, or list of DataFrame
```
Comment (datapythonista) [inline pandas/core/frame.py]: Do you mind removing the `with name field set` from the types, and have it in the description instead?

File: pandas/core/frame.py [hunk 3]
```diff
@@ -6449,31 +6451,31 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
             Index should be similar to one of the columns in this one. If a
             Series is passed, its name attribute must be set, and that will be
-            used as the column name in the resulting joined DataFrame
+            used as the column name in the resulting joined DataFrame.
         on : name, tuple/list of names, or array-like
```
Comment (datapythonista) [inline pandas/core/frame.py]: In the types, we try to have just the Python (or numpy...) types, and with a specific format that we can at some point parse and validate. Could you replace it by something like `str, list of str or array-like` please?

File: pandas/core/frame.py [hunk 4]
```diff
@@ -6449,31 +6451,31 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
               on is specified) with other frame's index, preserving the order
-              of the calling's one
+              of the calling's one.
         lsuffix : string
-            Suffix to use from left frame's overlapping columns
+            Suffix to use from left frame's overlapping columns.
         rsuffix : string
-            Suffix to use from right frame's overlapping columns
+            Suffix to use from right frame's overlapping columns.
         sort : boolean, default False
```
Comment (datapythonista) [inline pandas/core/frame.py]: Can you replace the type `string` by `str` and `boolean` by `bool`

File: pandas/core/frame.py [hunk 5]
```diff
@@ -6485,70 +6487,67 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
 
         Examples
         --------
+        >>> import pandas as pd
+
```
Comment (datapythonista) [inline pandas/core/frame.py]: I know this is being reported by flake8, but so far we're not adding them, and a decision of what to do in these cases needs to be made in #22900 first.

Can you remove this for now, and ignore the error from flake8 please?

File: pandas/core/frame.py [hunk 6]
```diff
@@ -6485,70 +6487,67 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
         >>> caller = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3', 'K4', 'K5'],
```
Comment (datapythonista) [inline pandas/core/frame.py]: do you mind replacing the name `caller` by `df`? we use this convention everywhere, I think it makes sense here too.

File: pandas/core/frame.py [hunk 7]
```diff
@@ -6485,70 +6487,67 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
+          key   A    B
+        0  K0  A0   B0
+        1  K1  A1   B1
+        2  K2  A2   B2
+        3  K3  A3  NaN
+        4  K4  A4  NaN
+        5  K5  A5  NaN
 
         See also
         --------
```
Comment (datapythonista) [inline pandas/core/frame.py]: Can you move the `See Also` section before the examples, use a capital `A` for `Also`, and add a period at the end of the `merge` description.

**Review by datapythonista — COMMENTED** [review]
Great changes, thanks a lot @JustinZhengBC. I added some comments mainly about the formatting of orignial the docstring. If you can fix those too, that would be great.

**Comment by JustinZhengBC** [comment]
Alright, I believe I have taken care of all the other issues now.

File: pandas/core/frame.py [hunk 8]
```diff
@@ -6440,40 +6440,42 @@ def append(self, other, ignore_index=False,
...
         Parameters
         ----------
-        other : DataFrame, Series with name field set, or list of DataFrame
+        other : DataFrame, Series, or list of DataFrame
             Index should be similar to one of the columns in this one. If a
             Series is passed, its name attribute must be set, and that will be
-            used as the column name in the resulting joined DataFrame
-        on : name, tuple/list of names, or array-like
+            used as the column name in the resulting joined DataFrame.
+        on : str, list of str, or array-like
```
Comment (datapythonista) [inline pandas/core/frame.py]: can you add that it's optional: `on : str, list of str, or array-like, optional`

File: pandas/core/frame.py [hunk 9]
```diff
@@ -6440,40 +6440,42 @@ def append(self, other, ignore_index=False,
...
-              of the calling's one
-        lsuffix : string
-            Suffix to use from left frame's overlapping columns
-        rsuffix : string
-            Suffix to use from right frame's overlapping columns
-        sort : boolean, default False
+              of the calling's one.
+        lsuffix : str
+            Suffix to use from left frame's overlapping columns.
+        rsuffix : str
```
Comment (datapythonista) [inline pandas/core/frame.py]: can you add the default: `str, default ''`

File: pandas/core/frame.py [hunk 10]
```diff
@@ -6483,76 +6485,72 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
         Support for specifying index levels as the `on` parameter was added
         in version 0.23.0
```
Comment (datapythonista) [inline pandas/core/frame.py]: Do you mind capitalizing the first letter of each sentence, and ending them in period in the Notes. For the First paragraph, if you can add backticks around the parameters `on`, `lsuffix`... that would be great too. So: "Options \`on\`, \`lsuffix\` and \`rsuffix\` are not supported [...]"

File: pandas/core/frame.py [hunk 11]
```diff
@@ -6483,76 +6485,72 @@ def join(self, other, on=None, how='left', lsuffix='', rsuffix='',
...
+          key   A    B
+        0  K0  A0   B0
+        1  K1  A1   B1
+        2  K2  A2   B2
+        3  K3  A3  NaN
+        4  K4  A4  NaN
+        5  K5  A5  NaN
 
         Returns
         -------
```
Comment (datapythonista) [inline pandas/core/frame.py]: Sorry, I missed before that the Returns was here. Can you move it after the parameters, and use this format:
```
Returns
-------
DataFrame
    A short description on what is being returned.
```

**Review by datapythonista — COMMENTED** [review]
Looks great, just couple of small things that I didn't see in the previous review, and I think we can merge this. Thanks!

**Comment by JustinZhengBC** [comment]
I fixed those issues. I also added some backticks around parameter names in other sections, as that was what other docstrings seemed to be doing.

File: pandas/core/frame.py [hunk 12]
```diff
@@ -6440,123 +6440,122 @@ def append(self, other, ignore_index=False,
...
+        K0   A0   B0
+        K1   A1   B1
+        K2   A2   B2
+        K3   A3  NaN
+        K4   A4  NaN
+        K5   A5  NaN
+
+        Another option to join using the key columns is to use the `on`
+        parameter. DataFrame.join always uses `other`'s index but we can use
+        any column in ``df``. This method preserves the original DataFrame's
```
Comment (datapythonista) [inline pandas/core/frame.py]: The double backticks is for code, and the single backticks for objects that are being referenced. We don't have an exact convention for things like the parameters or the variables. Can you leave them with single backticks for now, as it's inconsistent to have df with double, and the rest like other with single. Other than that, the PR looks great.

**Comment by JustinZhengBC** [comment]
I changed the backticks around the df's to be single again.

**Review by datapythonista — APPROVED** [review]
Looks great, thanks for the PR @JustinZhengBC 

We'll merge it once the CI checks are on green. Hope to see more of your contributions.

**Comment by datapythonista** [comment]
Superseded by #23471

**Pull request closed without merging**